### PR TITLE
Block Bindings: Improve how the context needed by sources is extended in the editor

### PR DESCRIPTION
**Pull request by SantosGuillamot**
## What?
Stealing the code from Ella in [this other pull request](https://github.com/WordPress/gutenberg/pull/60826).

This pull request uses the code from https://github.com/WordPress/gutenberg/pull/63470  to reuse and merge the `usesContext` defined by the source in the server.

There are a few things happening:
* By default, it reuses the `usesContext` defined by the source in the server registration.
* It adds a new `usesContext` property to the client source registration to let sources extend that if needed.
* It merges both and removes duplicates.
* It changes the way bindings access the context to ensure it is only available when an attribute is connected to that specific source.

## Why?
Block bindings shouldn't have to affect the block registration and the context can be accessed just in time.

Apart from that, it should reuse the server definition to avoid duplication in server and client registrations.

## Testing Instructions
I added some automated tests to ensure everything works as expected: [link](https://github.com/WordPress/gutenberg/compare/try/register-sources-from-the-server...try/improve-how-bindings-context-is-extended-in-the-editor?expand=1#diff-d5aba57655b6c1abcbd8f937a61ad98110057a6ae09f2db596811dca31c54030).

### CODE REVIEW

File: packages/blocks/src/api/test/registration.js [hunk 1]
```diff
@@ -1528,6 +1528,74 @@ describe( 'blocks', () => {
...
+				name: 'core/testing',
+				label: 'testing',
+				usesContext: 'should be an array',
+			} );
+			expect( console ).toHaveWarnedWith(
+				'Block bindings source usesContext must be an array.'
+			);
+		} );
+
+		it( 'should add usesContext when only defined in the client', () => {
```
Comment (artemiomorales) [inline packages/blocks/src/api/test/registration.js]: I believe this should read, "should add usesContext when only defined in the server".

File: packages/blocks/src/api/test/registration.js [hunk 2]
```diff
@@ -1528,6 +1528,74 @@ describe( 'blocks', () => {
...
+			registerBlockBindingsSource( {
+				name: 'core/testing',
+				getValue: () => 'value',
+			} );
+			const source = getBlockBindingsSource( 'core/testing' );
+			unregisterBlockBindingsSource( 'core/testing' );
+			expect( source.usesContext ).toEqual( [ 'postId', 'postType' ] );
+		} );
+
+		it( 'should keep usesContext when it is not defined in the client', () => {
```
Comment (artemiomorales) [inline packages/blocks/src/api/test/registration.js]: And I believe this one should read, "should keep usesContext when only defined in the client".

**Review by artemiomorales — APPROVED** [review]
Aside from the two descriptions for the tests below, this looks good to me 👍 I believe this is good to merge if those are addressed.

File: packages/blocks/src/api/test/registration.js [hunk 3]
```diff
@@ -1528,6 +1528,74 @@ describe( 'blocks', () => {
...
+			);
+		} );
+
+		it( 'should add usesContext when only defined in the client', () => {
+			// Simulate server bootstrap.
+			registerBlockBindingsSource( {
+				name: 'core/testing',
+				label: 'testing',
+				usesContext: [ 'postId', 'postType' ],
+			} );
```
Comment (artemiomorales) [inline packages/blocks/src/api/test/registration.js]: I see you're using `addBootstrappedBlockBindingsSource()` to simulate the server behavior in [a related PR](https://github.com/WordPress/gutenberg/pull/63797/files#). Should we be taking that approach in this PR as well? It seems like that would more closely resemble the server behavior.

Comment (SantosGuillamot) [reply packages/blocks/src/api/test/registration.js]: That makes sense 🙂 I'll make that change.

File: packages/blocks/src/api/test/registration.js (hunk 2 above)
Comment (SantosGuillamot) [reply packages/blocks/src/api/test/registration.js]: I will change the text.

File: packages/blocks/src/api/test/registration.js (hunk 3 above)
Comment (SantosGuillamot) [reply packages/blocks/src/api/test/registration.js]: I made the change as part of [this commit](https://github.com/WordPress/gutenberg/pull/63513/commits/1d7160491a65c402a9b2ca5da6d4c1a87ee0b8c7).

File: packages/blocks/src/api/test/registration.js (hunk 2 above)
Comment (SantosGuillamot) [reply packages/blocks/src/api/test/registration.js]: I made the change as part of [this commit](https://github.com/WordPress/gutenberg/pull/63513/commits/1d7160491a65c402a9b2ca5da6d4c1a87ee0b8c7).

File: packages/blocks/src/api/test/registration.js (hunk 1 above)
Comment (SantosGuillamot) [reply packages/blocks/src/api/test/registration.js]: I made the change as part of [this commit](https://github.com/WordPress/gutenberg/pull/63513/commits/1d7160491a65c402a9b2ca5da6d4c1a87ee0b8c7).

File: packages/blocks/src/api/test/registration.js (hunk 3 above)
Comment (gziolo) [reply packages/blocks/src/api/test/registration.js]: Nice, this improves the readability of the test 👍🏻

**Pull request merged**